### ISSUE TITLE
Some questions about deepwave

### ISSUE THREAD

**Issue by guoketing** [ar4/deepwave#32]
Dear Alan,

I'm very glad to get the so wonderful codes. I try to implement other PML methods some days ago, but the PML method in deepwave have good enough result. So I give up this idea. I have read most codes in the deepwave. I have same questions about it.  
question 1,
Is that possible to extend to wave case from acoustic case? I mean using velocity-stress form of the elastodynamic equations to generate seismic data and then we can inverse vp,vs and ρ using deepwave.
question 2,
Is the deepwave using RNN? I have read the paper that you wrote at 2018 "Seismic Full-Waveform Inversion Using Deep Learning Tools and Techniques". In this paper, you believe conventional seismic FWI can be constructed as a RNN, but I can't found any codes in deepwave about RNN. So, is that true that the forward function of finite difference is the process of RNN?

**Comment by ar4** [comment]
Dear Keting,

I am glad that the code was clear enough to be understood by someone else, and that you agree that the PML works adequately well.

It certainly should be possible to extend the code to work with any other finite difference wave propagator. When I wrote Deepwave I originally planned to do that, and so designed it in a way that I hoped would make it easy to add other propagators, but I moved onto other projects after doing the scalar wave equation case and haven't had time to add any more since then.

You are correct that Deepwave does not use PyTorch's RNN to perform the finite difference steps. The effect will be the same, but by manually controlling the finite difference forward and adjoint calculations I am able to get better computational performance and reduce the memory cost. Deepwave does still use the fact that wave propagation is a differentiable operation just like others used in deep learning networks, and uses this to integrate it into PyTorch. Wrapping the compiled finite difference code as a PyTorch module provides the advantage of being able to include wave propagation in a series of differentiable operations and easily forward and backpropagate through all of them, for example to add a fancy objective function to the output of wave propagation and let PyTorch's automatic differentiation work out how to backpropagate through the combination of both operations.

**Comment by guoketing** [comment]
Dear Alan,

Thank you for your reply, I have the idea that maybe I can extent the deepwave using other finite difference wave propagator and improve the deepwave module. I will try to find the way to extend to wave case. I appreciate you helping me so much. If there is anything I can help you, I will try my best to do it.

**Comment by ar4** [comment]
Dear Keting,

Extension to the elastic case would be nice. I imagine that a good way to
do it would be to create a new package that has Deepwave as a dependency.
You could then create a new propagator that inherits from Deepwave's
propagator (`class Propagator(deepwave.base.propagator.Propagator)`, as in
`scalar.py`). Let me know if you encounter problems.
